ClassicLootManager/Modules/Rosters/RosterManager/ExpressionParser.lua

```diff
@@ -0,0 +1,425 @@
+    for word in expression:gmatch(nonOperatorParenthesisOrNumber) do
+        if not variableValues[word] then
+            errors[#errors + 1] = errorMessage
+            return
+        end
+    end
+end
+
+local function verifyParenthesisAreBalanced(expression, errors)
+    local errorMessage = CLM.L["Expression has unbalanced parentheses"]
+
+    local nestDepth = 0
+    for i = 1, #expression do
+        local c = expression:sub(i, i)
+        if c == '(' then
+            nestDepth = nestDepth + 1
+        elseif c == ')' then
+            nestDepth = nestDepth - 1
+        end
+        if nestDepth < 0 then
+            errors[#errors + 1] = errorMessage
+            return
+        end
+    end
+    if nestDepth ~= 0 then
+        errors[#errors + 1] = errorMessage
+    end
+end
+
+local function tokenize(input)
+    local tokens = {}
+    local currentToken = ""
+    local index = 1;
+    for char in input:gmatch(".") do
+        if char:match("[%d%.]") or (char:match("-") and not currentToken:match("[%d]$")) then
+            -- gathering token parts
+            currentToken = currentToken .. char
+        else
+            -- write current token
+            currentToken = currentToken:gsub("%-%-", "") -- correction for negative numbers
+            tokens[index] = currentToken
+            currentToken = ""
+            index = index + 1
+
+            -- write new token
+            tokens[index] = char
+            index = index + 1
+        end
+    end
+    if currentToken ~= "" then
+        tokens[index] = currentToken
+    end
+    return tokens
+end
+
+local function processOperator(tokens, process)
+    for i, token in ipairs(tokens) do
+        if process[token] then
+            local j = i - 1
+            while tokens[j] == "" do
+                j = j - 1
+            end
+            local left = tonumber(tokens[j])
+            local right = tonumber(tokens[i + 1])
+            local result = process[token](left, right)
+            tokens[j] = result
+            tokens[i] = ""
+            tokens[i + 1] = ""
+        end
+    end
+end
+
+local function solve(eq)
+
+    while eq:find("%b()") do
+        eq = eq:gsub("%b()", function(block)
+            local inner = block:sub(2, -2)
+            return tostring(solve(inner))
+        end)
+    end
+
+    local tokens = tokenize(eq)
+    processOperator(tokens, {
+        ["^"] = function(left, right) return left ^ right end
+    })
+    processOperator(tokens, {
+        ["*"] = function(left, right) return left * right end,
+        ["/"] = function(left, right) return left / right end
+    })
+    processOperator(tokens, {
+        ["+"] = function(left, right) return left + right end,
+        ["-"] = function(left, right) return left - right end
+    })
+
+    return tonumber(tokens[1])
+end
+
+local function replaceVariablesWithValues(expression, variables)
+    -- Sort the variables by length to avoid replacing substrings
```

Comment:
I understand the need of doing that in a generic solution, but this is redundant for CLM implementation because you should guarantee the variables are sorted properly when passing them